spring-boot/admin/src/main/java/eu/coatrack/admin/service/OAuthUserDetailsService.java

```diff
@@ -0,0 +1,115 @@
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.List;
+
+@Service
+public class OAuthUserDetailsService {
+
+    private static final String GITHUB_API_EMAIL = "https://api.github.com/user/emails";
+
+    private final OAuth2AuthorizedClientService clientService;
+    private final RestTemplate restTemplate;
+
+    @Autowired
+    public OAuthUserDetailsService(OAuth2AuthorizedClientService clientService,
+                                   RestTemplate restTemplate) {
+        this.clientService = clientService;
+        this.restTemplate = restTemplate;
+    }
+
+    private OAuth2User getLoggedInUser() {
+        return (OAuth2User) SecurityContextHolder
+                .getContext()
+                .getAuthentication()
+                .getPrincipal();
+    }
+
+    public String getLoginNameFromLoggedInUser() {
+        return getLoggedInUser().getAttribute("login");
+    }
+
+    public String getNameFromLoggedInUser() {
+        return getLoggedInUser().getAttribute("name");
+    }
+
+    public String getCompanyFromLoggedInUser() {
+        return getLoggedInUser().getAttribute("company");
+    }
+
+    public String getEmailFromLoggedInUser() throws JsonProcessingException {
+        String email = getLoggedInUser().getAttribute("email");
+
+        if (email == null || email.isEmpty()) {
+            email = getPrimaryEmailFromLoggedInUser();
+        }
+        return email;
+    }
+
+    private List<GithubEmail> getEmailsListFromGithub() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("Authorization", "token " + getTokenFromLoggedInUser());
+        HttpEntity<String> githubRequest = new HttpEntity(headers);
+
+        ResponseEntity<List<GithubEmail>> emailsListFromGithub = restTemplate.exchange(GITHUB_API_EMAIL, HttpMethod.GET,
+                githubRequest, new ParameterizedTypeReference<List<GithubEmail>>() {});
+        return emailsListFromGithub.getBody();
+    }
+
+    private String getPrimaryEmailFromLoggedInUser() {
+
+        return getEmailsListFromGithub()
+                .stream()
+                .filter(email -> (email.isMailAddressVerified() && email.isUsersPrimaryMailAddress()))
+                .map(GithubEmail::getEmail)
+                .findFirst()
+                .orElse(null);
```

Comment:
See comment above referring to the return of null.

Reply:
Yes, so I requested an example from you of how this optional would work